### PR TITLE
KAFKA-7250: switch scala transform to TransformSupplier

### PR DESCRIPTION
**Pull request by vvcephei**
#5468 introduced a breaking API change that was actually avoidable. This PR re-introduces the old API as deprecated and alters the API introduced by #5468 to be consistent with the other methods

* also, fixed misc syntax problems

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    override def transform(key: String, value: Long): (Int, Double) = {
+      transforms = transforms + 1
+      (transforms, transforms)
+    }
+
+    override def close(): Unit = ???
+  }
+
+  private class MockJStream extends JStream[String, Long] {
```
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala]: I tried to use EasyMock, but the capture wouldn't work; I think because of the varargs parameter.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala [hunk 2]
```diff
@@ -207,17 +204,20 @@ class TopologyTest extends JUnitSuite {
...
                 override def init(context: ProcessorContext): Unit = Unit
+
                 override def transform(key: String, value: String): KeyValue[String, String] =
                   new KeyValue(key, value.toLowerCase)
+
                 override def close(): Unit = Unit
-            }
-          )
+              }
+          })
```
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:220]: I also compiled with scala 2.12 to verify that the old code works via SAM conversion. I don't think we can explicitly test with it in this project, though, until we update to 2.12.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 3]
```diff
@@ -295,9 +297,42 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * @return a [[KStream]] that contains more or less records with new key and value (possibly of different type)
    * @see `org.apache.kafka.streams.kstream.KStream#transform`
    */
-  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, KeyValue[K1, V1]],
+  def transform[K1, V1](transformerSupplier: TransformerSupplier[K, V, KeyValue[K1, V1]],
```
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:299]: Question: should we reform the process method to also expect a `ProcessorSupplier`? It shouldn't break anyone on 2.12+ due to SAM conversion.

**Comment by vvcephei** [comment]
@guozhangwang @mjsax : can you take a look at this? thanks.

@mdziemianko : We had some tests fail in another project due to the API breakage. I had an idea for how to preserve the pre-existing API while also fixing the bug you fixed. Do you mind sharing your thoughts on this PR?

File: build.gradle [hunk 4]
```diff
@@ -709,7 +709,6 @@ project(':core') {
                                'genAdminClientConfigDocs', 'genProducerConfigDocs', 'genConsumerConfigDocs',
                                'genKafkaConfigDocs', 'genTopicConfigDocs',
                                ':connect:runtime:genConnectConfigDocs', ':connect:runtime:genConnectTransformationDocs',
-                               ':connect:runtime:genSinkConnectorConfigDocs', ':connect:runtime:genSourceConnectorConfigDocs',
```
Comment (guozhangwang) [inline build.gradle]: Are these change intentional? Or maybe you need to rebase from latest trunk..

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 5]
```diff
@@ -176,15 +176,17 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
...
+   * Creates an array of `KStream` from this stream by branching the records in the original stream based on
    * the supplied predicates.
    *
    * @param predicates the ordered list of functions that return a Boolean
    * @return multiple distinct substreams of this [[KStream]]
    * @see `org.apache.kafka.streams.kstream.KStream#branch`
    */
-  def branch(predicates: (K, V) => Boolean*): Array[KStream[K, V]] =
+  //noinspection ScalaUnnecessaryParentheses
+  def branch(predicates: ((K, V) => Boolean)*): Array[KStream[K, V]] = {
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: Why this change?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 6]
```diff
@@ -211,7 +213,7 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
    * }}}
    *
    * @param topic the topic name
-   * @param (implicit) produced the instance of Produced that gives the serdes and `StreamPartitioner`
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: Why removing `implicit`?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 7]
```diff
@@ -295,9 +297,42 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
...
+    */
+  @deprecated("Use transform(TransformerSupplier,String) instead.", "2.0")
+  def transform[K1, V1](transformer: => Transformer[K, V, (K1, V1)], stateStoreNames: String*): KStream[K1, V1] = {
+    //noinspection ConvertExpressionToSAM // due to 2.11
+    val transformerSupplierJ: TransformerSupplier[K, V, KeyValue[K1, V1]] =
+      new TransformerSupplier[K, V, KeyValue[K1, V1]] {
+        override def get(): Transformer[K, V, KeyValue[K1, V1]] =
+          new Transformer[K, V, KeyValue[K1, V1]] {
+            override def transform(key: K, value: V): KeyValue[K1, V1] =
+              transformer.transform(key, value) match {
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: We should add in the document that users should not use a transformer instance as the parameter.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    val transformer1 = supplier.get()
+    val transformer2 = supplier.get()
+
+    // we should get two different instances
+    Assert.assertEquals(new KeyValue(1, 1.0), transformer1.transform("A", 1L))
+    Assert.assertEquals(new KeyValue(1, 1.0), transformer2.transform("A", 1L))
+  }
+
+  // demonstrating the condition under which the deprecated method yields incorrect behavior
+  @Test def deprecatedTransformShouldDoesntCreateATrueSupplierWhenPassedAnInstance(): Unit = {
```
Comment (guozhangwang) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala]: nit: `..ShouldNot..`

**Review by guozhangwang — COMMENTED** [review]
The proposed change lgtm. As for the deprecated API, if people feel that directly removing it is too harsh we can leave it, though personally I'd prefer just remove it with the new API that would not potentially leave buggy holes for users.

Have you validate this change with splotlessCheck?

File: build.gradle (hunk 4 above)
Comment (vvcephei) [reply build.gradle]: woah! it looks like I need to rebase.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 5 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: The type checker in IDEA was spazzing out and treating `predicates`'s type as `Seq[(K,V) => Any]`.
This was the only way I could get it to shut up.

I actually couldn't compile this class in IDEA without this change.

Funny enough, it fails without the parens and then complains that the parens are unnecessary once you add them.

Can you check in your IDEA that you get no reported error and that you can run the tests I added from IDEA? If so, maybe I have the ide misconfigured.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 6 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: It's a scaladoc syntax error.

Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: We could add it after `produced`, but AFAICT, the param name must immediately follow `@param`

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala (hunk 8 above)
Comment (vvcephei) [reply streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/KStreamTest.scala]: haha, fair enough!

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 6 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: unless my IDE is also wrong about this...

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 7 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: done.

**Comment by vvcephei** [comment]
@guozhangwang what do you think about this:

> Question: should we reform the process method to also expect a ProcessorSupplier? It shouldn't break anyone on 2.12+ due to SAM conversion.

**Comment by mdziemianko** [comment]
LGTM too.

I was hoping to have time to go over the whole DSL and make it more consistent as each function looks different:) I chose the generator function because it is similar to how map, join etc work. The transformValue are the only ones that accept suppliers as far as I can tell from a brief look over the code.

I am of an opinion that this DSL looks a bit like PHP standard library... each signature is different. It goes not only for object vs generator vs supplier, but also how the serdes are accepted (especially in KTable), for example:
def filter(predicate: (K, V) => Boolean, materialized: Materialized[K, V, ByteArrayKeyValueStore]): KTable[K, V]
versus:
def groupBy[KR, VR](selector: (K, V) => (KR, VR))(implicit serialized: Serialized[KR, VR]): KTable[K, VR] 
or my favourite:
def join[VO, VR](other: KTable[K, VO])(joiner: (V, VO) => VR, materialized: Materialized[K, VR, ByteArrayKeyValueStore]): KTable[K, VR] 

The point being - pick one scheme, I am willing to spend some time making sure that everything adhere to this scheme...

I also think a buggy discontinued method should be just deleted rather than deprecated especially in a DSL that just come out and not many people are using it yet.

**Comment by guozhangwang** [comment]
@mdziemianko The reason of the inconsistency in KTable function signatures is for implicit conversions: we want to ease users of specifying those parameters that can be inferred by Scala anyways. And for different functions the parameters that can be implicit are a bit different.

@vvcephei For `process()`, if we change the signature would that break for users on Scala 2.11?

**Comment by vvcephei** [comment]
@mdziemianko Thanks for the review.

I think we would welcome a PR to standardize the scala DSL. Please do, if you find the time!

On the deprecate-vs-delete, I think I tend to err on the side of deprecating in general, but this may be a special case. Special factors might be:
* the extreme youth of this API
* the fact that the interface itself is buggy

Usually, we can deprecate an unfortunate method and put in some kind of correct workaround implementation, but in this case, there's nothing we can do if you pre-instantiate a transformer and pass it in. We can't even detect it.

The by-reference argument makes it slightly better, since it *might* really act like a factory, but there's no guarantee.

Maybe it is better to just delete it in this case... 

Or we could keep the method but throw an exception in the body explaining the problem. I just imagine bumping up to a new bugfix version of 2.0.x is going to fail to compile, and we won't get a chance to explain why such an extreme measure was necessary. Of course, it'll be in the release notes, but I'd feel better if it was "in the loop" for the dev.

**Comment by vvcephei** [comment]
@guozhangwang yes, it would, but we could add an implicit conversion to ease the transition.

Also note that process is now the only method in the whole API that takes `() => X` instead of `XSupplier`.

EDIT: I take that back; according to my local tests, the two methods can exist side-by-side, so we can keep the `()=>X` version and add a `XSupplier` version. So no one would be broken.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 6 above)
Comment (mjsax) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: You can build the scala docs and double check if in doubt -- if is works same a JavaDocs, I think John is correct.

**Comment by vvcephei** [comment]
@guozhangwang Since it looks fine to add the process variant taking a supplier, I've gone ahead and added it.

**Comment by mdziemianko** [comment]
@guozhangwang actually, I just converted quite a bit chunk of code from lightbend DSL wrapper and I have a feeling that compiler was able to infer more types and convert/wrap functions to suppliers on its own.  I had to add quite a few type annotations, mostly (i think) around areas where FunctionConversions kick in but I was not paying too much attention given the amount of code I had to convert on short notice...

Is there any good reason why certain functions such as aggregate or join have the parameters list split? not talking about implicits but about regular parameters split into multiple params list. Are you expecting a lot of partial applications?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 7 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: What do you think about just throwing an explanatory exception instead?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 3 above)
Comment (mjsax) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:299]: I don't think we should change `process` because it would require a KIP. We should rather fix the bug only. Maybe be can create a Jira for it though.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 9]
```diff
@@ -295,9 +296,45 @@ class KStream[K, V](val inner: KStreamJ[K, V]) {
...
+   * This method is deprecated because there is a danger of passing a specific Transformer instance, which would
+   * result in incorrect execution. It's highly recommended to us [[KStream.transform(TransformSupplier, String*)]] instead.
+   *
+   * @param transformer the `Transformer` instance
+   * @param stateStoreNames     the names of the state stores used by the processor
+   * @return a [[KStream]] that contains more or less records with new key and value (possibly of different type)
+   * @see [[org.apache.kafka.streams.kstream.KStream#transform]]
+   */
+  @deprecated("Use transform(TransformerSupplier,String) instead.", "2.0")
+  def transform[K1, V1](transformer: => Transformer[K, V, (K1, V1)], stateStoreNames: String*): KStream[K1, V1] = {
```
Comment (mjsax) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala]: I don't think we need to keep a deprecated method -- the old method allows to compile code, but it will fail during runtime. Thus, nobody can actually use the old method and thus there is no reason to keep it.

**Comment by mjsax** [comment]
@John: btw: the PR title should reflect the Jira. It's not MINOR. KAFKA-7250 -- also, please reference this PR in a comment on the JIRA. Thx.

**Comment by vvcephei** [comment]
@mjsax Thanks, I've renamed the PR, and I'll add a comment to the Jira. I have also created https://issues.apache.org/jira/browse/KAFKA-7267 with my thoughts about the process method.

**Comment by vvcephei** [comment]
@mdziemianko @guozhangwang @mjsax 

I've updated this PR to drop the broken transform method (just as #5468 did). I thought that we could keep it using a by-reference parameter, but that API still couldn't be guaranteed to be correct.

@mjsax also pointed out that the existing transform method has been unusable since it was introduced, which is proof that no one is using it and therefore, it's ok to break the API to fix the bug.

Nevertheless, I've decided to keep the PR just to clean up a few things in the Scala API and also to use a `TransformerSupplier` instead of a `() => Transformer`. This is really a 6-one-way-half-dozen-the-other kind of thing, but here's my thought:
* Both `KTable#transform` and `KStream#transformValues` take suppliers. The only other method that takes a function is `KStream#process`. Given that we have to change transform to something, I read this as a 2:1 vote in favor of supplier.
* users starting from scala 2.12 can pass a function to a method expecting a supplier, and SAM conversion will make it work, without even needing an implicit conversion. If we expect a function, though, there's no automatic conversion from supplier.

Note also that I previously had a test, which I've removed, since there's nothing left in this PR for it to test.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala (hunk 6 above)
Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:214]: Confirmed: when I generate the scaladoc from trunk, it just skips this parameter and only documents `topic`, but when I generate it from this branch, it documents both `topic` and `produced`.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala [hunk 10]
```diff
@@ -31,8 +31,8 @@ import scala.collection.JavaConverters._
 /**
  * Wraps the Java class [[org.apache.kafka.streams.kstream.KStream]] and delegates method calls to the underlying Java object.
  *
- * @param [K] Type of keys
- * @param [V] Type of values
```
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala:35]: This was also invalid scaladoc and caused the type parameters not to be documented at all.

**Comment by guozhangwang** [comment]
LGTM. Waiting for jenkins to pass.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.0 as well.
